core/src/main/scala/shapeless/ops/hlists.scala

```diff
@@ -1276,10 +1329,10 @@ object hlist {
       }
 
     // let (H :: T) ∪ M  =  H :: (T ∪ M) when H ∉ M
-    implicit def hlistUnion1[H, T <: HList, M <: HList]
+    implicit def hlistUnion1_[H, T <: HList, M <: HList]
       (implicit
        u: Union[T, M],
-       f: FilterNot.Aux[M, H, M]
+       f: NotContainsConstraint[M, H]
```

Comment:
I wonder if it's better to move `NotContainsConstraint[M, H]` before `Union[T. M]`. In theory it should save some unnecessary computations but in practice I couldn't see a difference.